clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -251,13 +252,15 @@ private void handleProduceResponse(ClientResponse response, Map<TopicPartition,
             // if we have a response, parse it
             if (response.hasResponse()) {
                 ProduceResponse produceResponse = new ProduceResponse(response.responseBody());
-                for (Map.Entry<TopicPartition, ProduceResponse.PartitionResponse> entry : produceResponse.responses()
-                                                                                                         .entrySet()) {
+                for (Map.Entry<TopicPartition, ProduceResponse.PartitionResponse> entry : produceResponse.responses().entrySet()) {
                     TopicPartition tp = entry.getKey();
                     ProduceResponse.PartitionResponse partResp = entry.getValue();
                     Errors error = Errors.forCode(partResp.errorCode);
                     RecordBatch batch = batches.get(tp);
-                    completeBatch(batch, error, partResp.baseOffset, correlationId, now);
+                    completeBatch(batch, error, partResp.baseOffset, partResp.timestamp, correlationId, now);
+                    if (error == Errors.NONE)
+                        metadata.recordTopicTimestampType(tp.topic(),
```

Comment:
I agree about avoiding broker side recompression. However, I still feel like we can achieve the same behavior with less changes. Let me know if I am missing something, but couldn't we just do the following:
1. Producer always sets timestamps (either producer client or KafkaProducer), as if timestamp type == CreateTime. For compressed message: timestamp of the outer message is set to the largest timestamp of the inner messages.
2. Broker will overwrite timestamps if type == LogAppendTime. In case of compressed message, it will overwrite only outer message timestamp, and let inner messages have "wrong" timestamps. If message ever gets uncompressed on broker: set all inner timestamps to outer timestamp if type == LogAppendTime;
3. When messages get uncompressed on consumer (or any case when we don't have timestamp type info): We know that outer timestamp is either max of inner timestamps (if create time) or outer timestamp was overwritten (and so inner timestamps are not valid anymore). We check if outer (compressed message) timestamp == max of inner message timestamps, and if false, set all inner message timestamps to outer timestamp. If there is ever a case that overwriting timestamp results in a timestamp == max of inner message timestamps, this means that producer and broker times are in sync, and inner message timestamps should be close enough to outer timestamp to care overwriting it.